File: src/test/java/com/gitlab/lae/intellij/jump/tree/TreeTest.java
package com.gitlab.lae.intellij.jump.tree;

import com.google.common.collect.ImmutableMap;
import junit.framework.TestCase;

import java.util.stream.Stream;

public final class TreeTest extends TestCase {

    public void testToTree() {

        Tree<Integer, Integer> actual = TreeNode.of(
                Stream.of(0, 1, 2, 3, 4, 5, 6, 7, 8, 9),
                2);

        Tree<Integer, Integer> expected = TreeNode.of(ImmutableMap.of(
                0, TreeNode.of(ImmutableMap.of(
                        0, TreeNode.of(ImmutableMap.of(
                                0, TreeNode.of(ImmutableMap.of(
                                        0, TreeLeaf.of(0),
                                        1, TreeLeaf.of(1)
                                )),
                                1, TreeNode.of(ImmutableMap.of(
                                        0, TreeLeaf.of(2),
                                        1, TreeLeaf.of(3)
                                ))
                        )),
                        1, TreeNode.of(ImmutableMap.of(
                                0, TreeNode.of(ImmutableMap.of(
                                        0, TreeLeaf.of(4),
                                        1, TreeLeaf.of(5)
                                )),
                                1, TreeNode.of(ImmutableMap.of(
                                        0, TreeLeaf.of(6),
                                        1, TreeLeaf.of(7)
                                ))
                        ))
                )),
                1, TreeNode.of(ImmutableMap.of(
                        0, TreeLeaf.of(8),
                        1, TreeLeaf.of(9)
                ))
        ));

        assertEquals(expected, actual);
    }

}

File: src/test/java/com/gitlab/lae/intellij/jump/TreeTest.kt
package com.gitlab.lae.intellij.jump

class TreeTest : junit.framework.TestCase() {

  fun `test to tree`() {
    val actual = treeOf(listOf(0, 1, 2, 3, 4, 5, 6, 7, 8, 9), 2)
    val expected =
        TreeNode(
            0 to
                TreeNode(
                    0 to
                        TreeNode(
                            0 to TreeNode(0 to TreeLeaf(0), 1 to TreeLeaf(1)),
                            1 to TreeNode(0 to TreeLeaf(2), 1 to TreeLeaf(3))),
                    1 to
                        TreeNode(
                            0 to TreeNode(0 to TreeLeaf(4), 1 to TreeLeaf(5)),
                            1 to TreeNode(0 to TreeLeaf(6), 1 to TreeLeaf(7)))),
            1 to TreeNode(listOf(0 to TreeLeaf(8), 1 to TreeLeaf(9))))
    assertEquals(expected, actual)
  }
}
